ledger/internal/prefetcher/prefetcher.go

```diff
@@ -283,21 +287,60 @@ func (p *accountPrefetcher) prefetch(ctx context.Context) {
 				loadAccountsAddAccountTask(&stxn.Txn.Receiver, task, accountTasks, queue)
 				loadAccountsAddAccountTask(&stxn.Txn.CloseRemainderTo, task, accountTasks, queue)
 			case protocol.AssetConfigTx:
+				loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.ConfigAsset), basics.AssetCreatable, task, resourceTasks, queue)
 			case protocol.AssetTransferTx:
+				if !stxn.Txn.AssetSender.IsZero() {
+					loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					loadAccountsAddResourceTask(&stxn.Txn.AssetSender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+				} else {
+					loadAccountsAddResourceTask(&stxn.Txn.Sender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					if stxn.Txn.AssetAmount == 0 && (stxn.Txn.AssetReceiver == stxn.Txn.Sender) {
+						// opt in
+						loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					}
+				}
+				if !stxn.Txn.AssetReceiver.IsZero() {
+					loadAccountsAddResourceTask(&stxn.Txn.AssetReceiver, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+				}
+				if !stxn.Txn.AssetCloseTo.IsZero() {
+					loadAccountsAddResourceTask(&stxn.Txn.AssetCloseTo, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
+				}
 			case protocol.AssetFreezeTx:
+				if !stxn.Txn.FreezeAccount.IsZero() {
+					loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.FreezeAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					loadAccountsAddResourceTask(&stxn.Txn.FreezeAccount, basics.CreatableIndex(stxn.Txn.FreezeAsset), basics.AssetCreatable, task, resourceTasks, queue)
+					loadAccountsAddAccountTask(&stxn.Txn.FreezeAccount, task, accountTasks, queue)
+				}
 			case protocol.ApplicationCallTx:
+				if stxn.Txn.ApplicationID != 0 {
+					// load the global - so that we'll have the program
+					loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.ApplicationID), basics.AppCreatable, task, resourceTasks, queue)
+					// load the local - so that we'll have the local state
+					// TODO: this is something we need to decide if we want to enable, since not
+					// every application call would use local storage.
+					if (stxn.Txn.ApplicationCallTxnFields.OnCompletion == transactions.OptInOC) ||
+						(stxn.Txn.ApplicationCallTxnFields.OnCompletion == transactions.CloseOutOC) ||
+						(stxn.Txn.ApplicationCallTxnFields.OnCompletion == transactions.ClearStateOC) {
+						loadAccountsAddResourceTask(&stxn.Txn.Sender, basics.CreatableIndex(stxn.Txn.ApplicationID), basics.AppCreatable, task, resourceTasks, queue)
+					}
+				}
+
+				// do not preload Txn.ForeignApps, Txn.ForeignAssets, Txn.Accounts
+				// since they might be non-used arbitrary values
```

Comment:
They _could_ be, but it seems like we should try to load them anyway, and fail gracefully.

There's even a case to be made for load the cross product - every holding for each account/asset pair, and local state / app pair. But that seems like a lot. I don't know.

I have this funny idea that we should execute the transaction with a special ledger that responds only from cache, and fakes answers to things that aren't in cache, but records what was requested. And we use _that_ to decide what to prefetch.

Reply:
Yes, I agree, there should be a cache factory in ledger for Round X that is populated assembly time by data from X-1, and used in validation.
I think @cce added this into future improvements list for lowering round time

Reply:
agreed, why not return to prefetching foreign arrays like we did before?

Reply:
this might be abused, so I'd prefer not to.